### PR TITLE
Adding a catalog entry for the bootstrap sample

### LINKED ISSUE #2384: Add BootstrapSamplingTransform to DataOperationsCatalog
As noted in #933, the `BootstrapSamplingTransform` isn't something that we want recorded in pipelines (e.g. having it executed on a test set). However, it is super useful to have in our library, so it would be nice to have a way to use it in user-facing code.

### PR DESCRIPTION
**Pull request by rogancarr**
This PR adds a catalog entry for the bootstrap sample and creates a sample demonstrating its use.

Fixes #2384

### CODE REVIEW

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 1]
```diff
@@ -24,12 +24,46 @@ internal DataOperationsCatalog(IHostEnvironment env)
...
+        /// <remarks>
+        /// This sampler is a streaming version of a bootstrap sampler. Instead of taking the whole dataset into memory and resampling,
+        /// <see cref="BootstrapSample"/> streams through the dataset and uses a Poisson(1) distribution to select rows to add to the sample.
+        /// The <paramref name="complement"/> parameter allows for the creation of a bootstap sample and complementary out-of-bag sample by
+        /// using the same <paramref name="seed"/>.
+        /// </remarks>
+        /// <param name="input">The input data.</param>
+        /// <param name="complement">Whether this is the out-of-bag sample, that is, all those rows that are not selected by the transform.
+        /// Can be used to create a complementary pair of samples by using the same seed.</param>
+        /// <param name="seed">The random seed. If unspecified random state will be instead derived from the environment.</param>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: >environment [](start = 109, length = 11)

Context? or <see cref="MlContext"/> ?

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 2]
```diff
@@ -24,12 +24,46 @@ internal DataOperationsCatalog(IHostEnvironment env)
...
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[BootstrapSample](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs)]
+        /// ]]>
+        /// </format>
+        /// </example>
+        public IDataView BootstrapSample(IDataView input,
+            bool complement = BootstrapSamplingTransformer.Defaults.Complement,
+            uint? seed = null,
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: I think the seed is a more fundamental option than complement... it seems like whenever you *might* want to make use of a non-default argument of complement, you would need to use seed, whereas I can see situations where one might want to use seed and not *neecessarily* using complement. This suggests to me, seed goes first.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
+﻿using System;
+using System.Linq;
+using Microsoft.Data.DataView;
+using Microsoft.ML.Data;
+
+namespace Microsoft.ML.Samples.Dynamic
+{
+    public class BootstrapSample
```
Comment (TomFinley) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: >public class BootstrapSample [](start = 4, length = 28)

Is this intended to be a static class? It has no modifier.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            var data = reader.Read(dataFile);
+
+            // Create a training set by taking a bootstrap sample. The sample has about the same number of rows as the input dataset
+            // (it's an approximate streaming bootstrap), but will only use about 63% unique rows per sample (i.e. 1-e^-1).
+            var train = mlContext.Data.BootstrapSample(data, seed: 632);
+            // And create a training set consisting of the complementary set of rows unused above by using the complement parameter
+            // with the same seed as before.
+            var test = mlContext.Data.BootstrapSample(data, complement: true, seed: 632);
+
+            // Create a learning pipeline
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: Good point. I'll just focus on the sampling and show with the example how it works.

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs (hunk 1 above)
Comment (rogancarr) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: Good catch!

File: src/Microsoft.ML.Data/Transforms/BootstrapSamplingTransformer.cs [hunk 5]
```diff
@@ -29,7 +29,7 @@ namespace Microsoft.ML.Transforms
     [BestFriend]
     internal sealed class BootstrapSamplingTransformer : FilterBase
     {
-        private static class Defaults
+        public static class Defaults
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply src/Microsoft.ML.Data/Transforms/BootstrapSamplingTransformer.cs]: Good eye.

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 6]
```diff
@@ -24,12 +24,46 @@ internal DataOperationsCatalog(IHostEnvironment env)
...
+            bool shuffleInput = BootstrapSamplingTransformer.Defaults.ShuffleInput,
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: I trimmed these out of the API.

Note that these options still exist in the base implementation, and the `BootstrapSample` filter *does* shuffle before and after resampling.

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs (hunk 2 above)
Comment (rogancarr) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: Done.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs (hunk 3 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: Let's keep it static.

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    public static class BootstrapSample
+    {
+        public static void Example()
+        {
+            // Create a new context for ML.NET operations. It can be used for exception tracking and logging, 
+            // as a catalog of available operations and as the source of randomness.
+            var mlContext = new MLContext();
+
+            // Get a small dataset as an IEnumerable and them read it as ML.NET's data type.
+            IEnumerable<SamplesUtils.DatasetUtils.BinaryLabelFloatFeatureVectorSample> enumerableOfData = SamplesUtils.DatasetUtils.GenerateBinaryLabelFloatFeatureVectorSamples(5);
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: >IEnumerable<SamplesUtils.DatasetUtils.BinaryLabelFloatFeatureVectorSample> [](start = 12, length = 74)

var?

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (TomFinley) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: Great. We will probably want to set these to be false, that is, do shuffling on neither end.

---
In reply to: [253351646](https://github.com/dotnet/machinelearning/pull/2387#discussion_r253351646) [](ancestors = 253351646)

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs (hunk 7 above)
Comment (rogancarr) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: The current style of these Samples is to use these really long names to be explicit about types....

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 8]
```diff
@@ -24,12 +24,42 @@ internal DataOperationsCatalog(IHostEnvironment env)
             Environment = env;
         }
 
+        /// <summary>
+        /// Take an approximate bootstrap sample of <paramref name="input"/>.
+        /// </summary>
+        /// <remarks>
+        /// This sampler is a streaming version of a bootstrap sampler. Instead of taking the whole dataset into memory and resampling,
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: Done.

File: src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs [hunk 9]
```diff
@@ -24,12 +24,42 @@ internal DataOperationsCatalog(IHostEnvironment env)
...
+        /// <see cref="BootstrapSample"/> streams through the dataset and uses a Poisson(1) distribution to select rows to add to the sample.
```
[the remark this answers is not in the document]
Comment (rogancarr) [reply src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs]: Done.

File: src/Microsoft.ML.Data/Transforms/BootstrapSamplingTransformer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (TomFinley) [reply src/Microsoft.ML.Data/Transforms/BootstrapSamplingTransformer.cs]: FYI I later realized that I had missed that the class above was internal already (which I did not notice, despite the fact that it's right on the line above this line), so the original change was actually correct, and nothing was being made public. Sorry @rogancarr.

**Pull request merged**